src/engraving/dom/fret.cpp

```diff
@@ -127,48 +133,79 @@ EngravingItem* FretDiagram::linkedClone()
-    int strings = static_cast<int>(s.size());
 
-    fd->setStrings(strings);
-    fd->setFrets(4);
-    fd->setPropertyFlags(Pid::FRET_STRINGS, PropertyFlags::UNSTYLED);
-    fd->setPropertyFlags(Pid::FRET_FRETS,   PropertyFlags::UNSTYLED);
-    int offset = 0;
-    int barreString = -1;
-    std::vector<std::pair<int, int> > dotsToAdd;
+    applyDiagramPattern(fd.get(), s);
 
-    for (int i = 0; i < strings; i++) {
-        Char c = s.at(i);
-        if (c == 'X' || c == 'O') {
-            FretMarkerType mt = (c == 'X' ? FretMarkerType::CROSS : FretMarkerType::CIRCLE);
-            fd->setMarker(i, mt);
-        } else if (c == '-' && barreString == -1) {
-            barreString = i;
-        } else {
-            int fret = c.digitValue();
-            if (fret != -1) {
-                dotsToAdd.push_back(std::make_pair(i, fret));
-                if (fret - 3 > 0 && offset < fret - 3) {
-                    offset = fret - 3;
+    return fd;
+}
+
+void FretDiagram::readHarmonyToDiagramFile(const muse::io::path_t& filePath)
+{
+    muse::io::File file(filePath);
+    if (!file.open()) {
+        LOGE() << file.errorString();
+        return;
+    }
+
+    XmlReader reader(&file);
+
+    while (reader.readNextStartElement()) {
+        if (reader.name() != "Data") {
+            break;
+        }
+        while (reader.readNextStartElement()) {
+            if (reader.name() == "HarmonyToDiagram") {
+                String harmony;
+                String diagram;
+                while (reader.readNextStartElement()) {
+                    if (reader.name() == "Harmony") {
+                        while (reader.readNextStartElement()) {
+                            if (reader.name() == "name") {
+                                harmony = reader.readText();
+                            } else {
+                                reader.unknown();
+                            }
+                        }
+                    } else if (reader.name() == "FretDiagram") {
+                        diagram = reader.readBody();
+                        reader.skipCurrentElement();
+                    } else {
+                        reader.unknown();
+                    }
+                }
+
+                if (!harmony.isEmpty() && !diagram.isEmpty()) {
+                    s_harmonyToDiagramMap.insert({ harmony, diagram });
                 }
+                harmony.clear();
+                diagram.clear();
+            } else {
+                reader.unknown();
             }
         }
     }
+}
 
-    if (offset > 0) {
-        fd->setFretOffset(offset);
-    }
+void FretDiagram::updateDiagram(const String& harmonyName)
+{
+    String harmonyNameLower = harmonyName.toLower();
 
-    for (std::pair<int, int> d : dotsToAdd) {
-        fd->setDot(d.first, d.second - offset, true);
-    }
+    auto it = std::find_if(s_harmonyToDiagramMap.cbegin(), s_harmonyToDiagramMap.cend(), [&harmonyNameLower](const auto& pair){
```

Comment:
This is essentially a linear algorithm, we don't use any benefits of std::map here. And it should be a constant complexity instead (without converting thousands of strings to lower case)